cmd/plugins/topology-aware/policy/pools.go

```diff
@@ -29,308 +29,249 @@ import (
+	dram, pmem, hbm = idset.NewIDSet(), idset.NewIDSet(), idset.NewIDSet()
+
+	for _, nodeID := range ids.SortedMembers() {
+		for _, closeID := range close[nodeID].SortedMembers() {
+			node := p.sys.Node(closeID)
+			switch node.GetMemoryType() {
+			case system.MemoryTypeDRAM:
+				dram.Add(closeID)
+			case system.MemoryTypePMEM:
+				pmem.Add(closeID)
+			case system.MemoryTypeHBM:
+				hbm.Add(closeID)
+			}
+		}
 	}
 
-	return assigned
+	return dram, pmem, hbm
 }
 
-// checkHWTopology verifies our otherwise implicit assumptions about the HW.
-func (p *policy) checkHWTopology() (bool, error) {
-	// NUMA nodes (memory controllers) should not be shared by multiple sockets.
-	socketNodes := map[idset.ID]cpuset.CPUSet{}
-	for _, socketID := range p.sys.PackageIDs() {
-		pkg := p.sys.Package(socketID)
-		socketNodes[socketID] = system.CPUSetFromIDSet(idset.NewIDSet(pkg.NodeIDs()...))
-	}
-	for id1, nodes1 := range socketNodes {
-		for id2, nodes2 := range socketNodes {
-			if id1 == id2 {
+func (p *policy) getCloseMemsWithCpu() map[idset.ID]idset.IDSet {
+	close := map[idset.ID]idset.IDSet{}
+
+	for _, nodeID := range p.sys.NodeIDs() {
+		node := p.sys.Node(nodeID)
+		if !node.CPUSet().IsEmpty() {
+			continue
+		}
+		close[nodeID] = idset.NewIDSet()
+		dist := -1
+		for id, d := range node.Distance() {
+			switch {
+			case id == nodeID:
 				continue
-			}
-			if shared := nodes1.Intersection(nodes2); !shared.IsEmpty() {
-				log.Error("can't handle HW topology: sockets #%v, #%v share NUMA node(s) #%s",
-					id1, id2, shared.String())
-				return false, policyError("unhandled HW topology: sockets #%v, #%v share NUMA node(s) #%s",
-					id1, id2, shared.String())
+			case dist == -1 || d < dist:
+				dist = d
+				close[nodeID] = idset.NewIDSet(id)
+			case d == dist:
+				close[nodeID].Add(id)
 			}
 		}
 	}
 
-	// NUMA nodes (memory controllers) should not be shared by multiple dies.
-	for _, socketID := range p.sys.PackageIDs() {
-		pkg := p.sys.Package(socketID)
-		for _, id1 := range pkg.DieIDs() {
-			nodes1 := idset.NewIDSet(pkg.DieNodeIDs(id1)...)
-			for _, id2 := range pkg.DieIDs() {
-				if id1 == id2 {
-					continue
-				}
-				nodes2 := idset.NewIDSet(pkg.DieNodeIDs(id2)...)
-				if shared := system.CPUSetFromIDSet(nodes1).Intersection(system.CPUSetFromIDSet(nodes2)); !shared.IsEmpty() {
-					log.Error("will ignore dies: "+
-						"socket #%v, dies #%v,%v share NUMA node(s) #%s",
-						socketID, id1, id2, shared.String())
-					return true, nil
-				}
+	reverse := map[idset.ID]idset.IDSet{}
+	for id, closeIDs := range close {
+		for _, closeID := range closeIDs.SortedMembers() {
+			if reverse[closeID] == nil {
+				reverse[closeID] = idset.NewIDSet(id)
+			} else {
+				reverse[closeID].Add(id)
 			}
 		}
 	}
 
+	log.Info("closest memory nodes with CPUs: %v", close)
+	log.Info("reverse closest memory nodes with CPUs: %v", reverse)
```

Comment:
I'm not sure about significant benefits of these logs, but I don't see any reason to block the PR due to this.

Reply:
The returned information directly controls how nodes without any CPU locality will get assigned to pools. These nodes are currently often PMEM or HBM, and I think in the future they could be CXL-attached memory as well. That's the reason why we log it, to provide some info related to their eventual assignment.

But I'll check and see if I can either make the logging more reasonable, or get rid of at least one of those messages.